subprojects/version-control/src/main/java/org/gradle/vcs/internal/DefaultVersionControlSystemFactory.java

```diff
@@ -16,17 +16,85 @@
 
 package org.gradle.vcs.internal;
 
+import org.gradle.cache.CacheRepository;
+import org.gradle.cache.FileLockManager;
+import org.gradle.cache.PersistentCache;
 import org.gradle.vcs.VersionControlSpec;
 import org.gradle.vcs.VersionControlSystem;
+import org.gradle.vcs.VersionRef;
 import org.gradle.vcs.git.internal.GitVersionControlSystem;
 
+import java.io.File;
+import java.util.Set;
+
+import static org.gradle.cache.internal.filelock.LockOptionsBuilder.mode;
+
 public class DefaultVersionControlSystemFactory implements VersionControlSystemFactory {
+    private final CacheRepository cacheRepository;
+
+    DefaultVersionControlSystemFactory(CacheRepository cacheRepository) {
+        this.cacheRepository = cacheRepository;
+    }
+
     @Override
     public VersionControlSystem create(VersionControlSpec spec) {
         // TODO: Register these mappings somewhere
+        VersionControlSystem vcs;
         if (spec instanceof DirectoryRepositorySpec) {
-            return new SimpleVersionControlSystem();
+            vcs = new SimpleVersionControlSystem();
+        } else {
+            vcs = new GitVersionControlSystem();
+        }
+        return new ThreadSafeVersionControlSystem(vcs, cacheRepository);
+    }
+
+    private static final class ThreadSafeVersionControlSystem implements VersionControlSystem {
+        private final VersionControlSystem delegate;
+        private final CacheRepository cacheRepository;
+
+        private ThreadSafeVersionControlSystem(VersionControlSystem delegate, CacheRepository cacheRepository) {
+            this.delegate = delegate;
+            this.cacheRepository = cacheRepository;
+        }
+
+        @Override
+        public Set<VersionRef> getAvailableVersions(VersionControlSpec spec) {
+            return delegate.getAvailableVersions(spec);
+        }
+
+        @Override
+        public File populate(final File versionDir, final VersionRef ref, final VersionControlSpec spec) {
+            PersistentCache cache = cacheRepository
+                .cache(versionDir)
+                .withLockOptions(mode(FileLockManager.LockMode.Exclusive))
+                .open();
+            DelegatePopulator populator = new DelegatePopulator(delegate, versionDir, ref, spec);
+            try {
+                cache.useCache(populator);
```

Comment:
You should use a `useCache(Factory<? extends T> action)` here instead of creating a `DelegatePopulator` and pulling the `workingDir` field out of it.  That would make `DelegatePopulator` a `Factory<File>` (I would maybe just call it `Populator`) or you can just inline the class here.